timemachine/md/builders.py

```diff
@@ -34,45 +129,103 @@ def build_protein_system(host_pdbfile: Union[app.PDBFile, str], protein_ff: str,
+        water_idxs = np.arange(host_coords.shape[0], solvated_host_coords.shape[0])
+        replace_clashy_waters(modeller, solvated_host_coords, box, water_idxs, mols, host_ff, water_ff)
+        solvated_host_coords = strip_units(modeller.positions)
+
+    num_water_atoms = solvated_host_coords.shape[0] - num_host_atoms
 
-    print("building a protein system with", nha, "protein atoms and", nwa, "water atoms")
+    assert modeller.getTopology().getNumAtoms() == solvated_host_coords.shape[0]
+
+    print("building a protein system with", num_host_atoms, "protein atoms and", num_water_atoms, "water atoms")
     solvated_host_system = host_ff.createSystem(
         modeller.topology, nonbondedMethod=app.NoCutoff, constraints=None, rigidWater=False
     )
 
-    return solvated_host_system, solvated_host_coords, box, modeller.topology, nwa
+    # Determine box from the system's coordinates
+    box = get_box_from_coords(solvated_host_coords)
+
+    return solvated_host_system, solvated_host_coords, box, modeller.topology, num_water_atoms
+
+
+def build_water_system(box_width: float, water_ff: str, mols: Optional[List[Chem.Mol]] = None):
+    """
+    Build a water system with a cubic box with each side of length box_width.
+
+    Parameters
+    ---------
+    box_width: float
+        The length of each size of the box
 
+    water_ff: str
+        The water forcefield name (excluding .xml) to parametrize the water with.
 
-def build_water_system(box_width, water_ff: str):
+    mols: optional list of mols
+        Molecules to be part of the system, will remove water molecules that clash with the mols.
+
+    Returns
+    -------
+    4-Tuple
+        OpenMM host system, coordinates, box, OpenMM topology
+    """
     ff = app.ForceField(f"{water_ff}.xml")
 
     # Create empty topology and coordinates.
     top = app.Topology()
     pos = unit.Quantity((), unit.angstroms)
-    m = app.Modeller(top, pos)
+    modeller = app.Modeller(top, pos)
+
+    box = np.eye(3, dtype=np.float64) * box_width
+    modeller.addSolvent(ff, boxSize=np.diag(box) * unit.nanometers, model=sanitize_water_ff(water_ff))
+
+    def get_host_coords():
+        host_coords = strip_units(modeller.positions)
+        # If mols provided, center waters such that the center is the mols centroid
+        # Done to avoid placing mols at the edges and moves the water coordinates to avoid
+        # changing the mol coordinates which are finalized downstream of the builder
+        if mols is not None and len(mols) > 0:
+            mol_coords = np.concatenate([get_romol_conf(mol) for mol in mols])
+            mols_centroid = np.mean(mol_coords, axis=0)
+            box_extents = (np.max(host_coords, axis=0) - np.min(host_coords, axis=0)) * 0.5
+            box_center = np.min(host_coords, axis=0) + box_extents
+            host_coords = host_coords - box_center + mols_centroid
+        return host_coords
```

Comment:
nit: this looks correct to me, but seems worth testing. Do you think it's worth moving it up to module scope and adding a test?

Reply:
Added a test in the builders in https://github.com/proteneer/timemachine/pull/1346/commits/d603cc0c9df8ba19f4391e3daad90185fdd80c5d

Verifies that if you pass ligands it changes the centroid to match the ligands and without passing the mols the centroid doesn't match.